elements/lisk-transactions/src/13_vote_transaction.ts

```diff
@@ -0,0 +1,450 @@
+				return 1;
+			}
+			if (diff < BigInt(0)) {
+				return -1;
+			}
+
+			return 0;
+		});
+		const errors = [];
+
+		for (const vote of assetCopy) {
+			const sender = await store.account.get(this.senderId);
+			const delegate = await store.account.get(vote.delegateAddress);
+
+			if (!delegate.username) {
+				errors.push(
+					new TransactionError(
+						'Voted delegate is not registered',
+						this.id,
+						'.asset.votes.delegateAddress',
+					),
+				);
+				continue;
+			}
+			if (vote.amount < BigInt(0)) {
+				const originalUpvoteIndex = sender.votes.findIndex(
+					senderVote => senderVote.delegateAddress === vote.delegateAddress,
+				);
+				if (originalUpvoteIndex < 0) {
+					errors.push(
+						new TransactionError(
+							'Cannot cast downvote to delegate who is not upvoted',
+							this.id,
+							'.asset.votes.delegateAddress',
+						),
+					);
+					continue;
+				}
+				sender.votes[originalUpvoteIndex].amount += vote.amount;
+				if (sender.votes[originalUpvoteIndex].amount < BigInt(0)) {
+					errors.push(
+						new TransactionError(
+							'Cannot downvote more than upvoted',
+							this.id,
+							'.asset.votes.amount',
+						),
+					);
+				}
+				// Delete entry when amount becomes 0
+				if (sender.votes[originalUpvoteIndex].amount === BigInt(0)) {
+					sender.votes = sender.votes.filter(
+						senderVote => senderVote.delegateAddress !== vote.delegateAddress,
+					);
+				}
+				// Create unlocking object
+				sender.unlocking.push({
+					delegateAddress: vote.delegateAddress,
+					amount: BigInt(-1) * vote.amount,
+					unvoteHeight: store.chain.lastBlockHeader.height + 1,
+				});
+				// Sort account.unlocking
+				sender.unlocking.sort((a, b) =>
+					a.delegateAddress.localeCompare(b.delegateAddress, 'en'),
+				);
+
+				// Unlocking object should not exceed maximum
+				if (sender.unlocking.length > MAX_UNLOCKING) {
+					errors.push(
+						new TransactionError(
+							`Cannot downvote which exceeds account.unlocking to have more than ${MAX_UNLOCKING}`,
+							this.id,
+							'.asset.votes',
+						),
+					);
+				}
+			} else {
+				// Upvote amount case
+				const originalUpvoteIndex = sender.votes.findIndex(
+					senderVote => senderVote.delegateAddress === vote.delegateAddress,
+				);
+				const index =
+					originalUpvoteIndex > -1 ? originalUpvoteIndex : sender.votes.length;
+				const upvote =
+					originalUpvoteIndex > -1
+						? sender.votes[originalUpvoteIndex]
+						: {
+								delegateAddress: vote.delegateAddress,
+								amount: BigInt(0),
+						  };
+				upvote.amount += vote.amount;
+				// Special case for postgres because maximum is int64 for bigint in postgres
+				if (upvote.amount > BigInt(MAX_INT64)) {
+					errors.push(
+						new TransactionError(
+							'Cannot upvote which exceeds int64',
+							this.id,
+							'.asset.votes.amount',
+						),
+					);
+				}
+				sender.balance -= vote.amount;
+				if (sender.balance < BigInt(0)) {
```

Comment:
should that be MIN_BALANCE? or are we checking for this somewhere else?

Reply:
this check is actually not necessary here. It is checked in the base transaction